hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java

```diff
@@ -204,12 +178,36 @@ private void scan() {
             log(invocationCount, backupTimeouts, invocationTimeouts);
         }
 
-        private boolean callIdMatches(long callId, Invocation invocation) {
-            return callId == invocation.op.getCallId();
-        }
+        /**
+         * The reason for the following if check is a workaround for the problem explained below.
+         *
+         * Problematic scenario :
+         * If an invocation with callId 1 is retried twice for any reason,
+         * two new innovations created and registered to invocation registry with callId’s 2 and 3 respectively.
+         * Both new invocations are sharing the same operation
+         * When one of the new invocations, say the one with callId 2 finishes, it de-registers itself from the
+         * invocation registry.
+         * When doing the de-registration it sets the shared operation’s callId to 0.
+         * After that when the invocation with the callId 3 completes, it tries to de-register itself from
+         * invocation registry
+         * but fails to do so since the invocation callId and the callId on the operation is not matching anymore
+         * When InvocationMonitor thread kicks in, it sees that there is an invocation in the registry,
+         * and asks whether invocation is finished or not.
+         * Even if the remote node replies with invocation is timed out,
+         * It can’t be de-registered from the registry because of aforementioned non-matching callId scenario.
+         *
+         * Workaround:
+         * When InvocationMonitor kicks in, it will do a check for invocations that are completed
+         * but their callId's are not matching with their operations. If any invocation found for that type,
+         * it is removed from the invocation registry.
+         */
+        private boolean duplicate(Invocation inv, long callId, Iterator iterator) {
```

Comment:
Maybe `isDuplicate()` is better to ease the read flow in `if (isDuplicate())`. Otherwise it sounds like you act on a successful duplication of something.